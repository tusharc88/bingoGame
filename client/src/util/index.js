export const generateBingoCardNumbers = matrixNum => {
  const numberOfRandomNumbers = matrixNum * matrixNum;
  const bingoCardNumbers = [];
  let counter = 0;

  while (counter < numberOfRandomNumbers) {
    const randomNumber = Math.floor(Math.random() * 100) + 1;
    if (bingoCardNumbers.indexOf(randomNumber) === -1) {
      bingoCardNumbers[counter] = randomNumber;
      counter += 1;
    }
  }

  return bingoCardNumbers;
};

export const generateUniqueRandomNumber = pastRandomNumbers => {
  if (pastRandomNumbers.length >= 100) return -1;
  while (true) {
    const randomNumber = Math.floor(Math.random() * 100) + 1;
    if (pastRandomNumbers.indexOf(randomNumber) === -1) return randomNumber;
  }
};

export const isBingo = (pastRandomNumbers, bingoCardNumbers) =>
  bingoCardNumbers.every(
    cardNumber => pastRandomNumbers.indexOf(cardNumber) > -1
  );
