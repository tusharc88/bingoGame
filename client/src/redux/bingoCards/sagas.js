import { takeEvery, select, put, all } from 'redux-saga/effects';
import { selectBingoCards } from './selectors';
import { populateCards } from './actions';
import { generateBingoCardNumbers } from '../../util/index';

export function* populateBingoCards() {
  const cards = yield select(selectBingoCards);
  const cardsWithNumbers = yield all(
    Object.keys(cards).map(cardNumber =>
      generateBingoCardNumbers(5, cardNumber)
    )
  );

  yield put(populateCards.success(cardsWithNumbers));
}

export function* watchPopulateCards() {
  yield takeEvery('POPULATE_CARDS_REQUEST', populateBingoCards);
}
