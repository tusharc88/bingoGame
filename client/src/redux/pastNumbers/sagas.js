import { takeEvery, call, put } from 'redux-saga/effects';
import { getNextNumber as getNextNumberAction } from './actions';
import { generateUniqueRandomNumber } from '../../util/index';

export function* getNextNumber(pastNumbers) {
  const uniqueRandomNumber = yield call(
    generateUniqueRandomNumber,
    pastNumbers
  );

  yield put(getNextNumberAction.success(uniqueRandomNumber));
}

export const callGetNextNumber = action => getNextNumber(action.payload);

export function* watchGetNextNumber() {
  yield takeEvery('NEXT_NUMBER_REQUEST', callGetNextNumber);
}
