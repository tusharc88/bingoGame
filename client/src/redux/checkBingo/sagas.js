import { takeEvery, select, put, call } from 'redux-saga/effects';
import { selectBingoCard } from '../bingoCards/selectors';
import { selectPastNumbers } from '../pastNumbers/selectors';
import { isBingo as isBingoAction } from './actions';
import { isBingo } from '../../util/index';

export function* checkIsBingo(cardNum) {
  const bingoCard = yield select(selectBingoCard(cardNum));
  const pastNumbers = yield select(selectPastNumbers);
  const isBingoResult = yield call(isBingo, pastNumbers, bingoCard);

  yield put(isBingoAction.success(isBingoResult));
}

export const callCheckBingo = action => checkIsBingo(action.payload);

export function* watchCheckBingo() {
  yield takeEvery('CHECK_BINGO_REQUEST', callCheckBingo);
}
