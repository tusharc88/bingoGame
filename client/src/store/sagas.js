import { all, fork } from 'redux-saga/effects';
import { watchPopulateCards } from '../redux/bingoCards/sagas';
import { watchCheckBingo } from '../redux/checkBingo/sagas';
import { watchGetNextNumber } from '../redux/pastNumbers/sagas';

export default function* rootSaga() {
  yield all([
    fork(watchPopulateCards),
    fork(watchCheckBingo),
    fork(watchGetNextNumber),
  ]);
}
