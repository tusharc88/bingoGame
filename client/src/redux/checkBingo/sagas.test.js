import { takeEvery, call } from 'redux-saga/effects';
import { callCheckBingo, checkIsBingo, watchCheckBingo } from './sagas';
import { isBingo } from '../../util/index';

test('checkIsBingo', () => {
  const generator = checkIsBingo(0);

  const dummyPastNumbers = [1, 2, 5, 7, 8, 21, 31, 3, 9, 91, 12, 29, 45, 30];
  const dummyFilledCard = [1, 5, 7, 8, 3, 9, 12, 29, 30];

  expect(generator.next().value).toMatchSnapshot();
  expect(generator.next(dummyFilledCard).value).toMatchSnapshot();
  expect(generator.next(dummyPastNumbers).value).toEqual(
    call(isBingo, dummyPastNumbers, dummyFilledCard)
  );
  expect(generator.next(true).value.PUT.action.payload).toEqual(true);
});

test('watchCheckBingo', () => {
  const generator = watchCheckBingo();
  expect(generator.next().value).toEqual(
    takeEvery('CHECK_BINGO_REQUEST', callCheckBingo)
  );
});
