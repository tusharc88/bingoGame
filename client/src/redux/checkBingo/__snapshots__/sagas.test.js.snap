// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkIsBingo 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`checkIsBingo 2`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;
